### PR TITLE
Added PathFollowing Functionality and odometry

### PR DESCRIPTION
**Pull request by noam987**

### CODE REVIEW

**Comment by noam987** [comment]
Note the pids have not been tuned. That will be done ASAP

**Comment by thetaspirit** [comment]
For all of the nav Commands, stick them in a folder called nav.  Put that folder in either the subsystems folder or the commands folder

File: code/src/main/java/frc/robot/commands/OdometryCommand.java [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+            if (currentPosIndex < m_path.length) {
+                m_xPIDController.setSetpoint(m_path[currentPosIndex].getTranslation().getX());
+                m_yPIDController.setSetpoint(m_path[currentPosIndex].getTranslation().getY());
+                m_thetaController.setGoal(m_path[currentPosIndex].getRotation().getRadians());
+                m_xPIDController.reset();
+                m_yPIDController.reset();
+
+            }
+
+        }
```
Comment (dotnetprofessional) [inline code/src/main/java/frc/robot/commands/OdometryCommand.java]: I don't understand why this needs to be outside the subsystem. I understand that you need something to keep polling the subsystem for position, however all the calculations seem like they can be done just as well in the subsystem with check for current x,y or simply a amIThereYet method. I'm most concerned that this is going to be hard to mock.

Is one of the reasons due to the amount of code needed in the subsystem? I maybe missing something.

File: code/src/main/java/frc/robot/commands/MoveToLocationCommand.java [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+import frc.robot.common.ILogger;
+import frc.robot.common.TraceableCommand;
+
+public class MoveToLocationCommand extends TraceableCommand {
+    private Pose2d[] path;
+    private IDrivetrainSubsystem m_subsystem;
+    private OdometryCommand m_command;
+
+    @Inject
+    public MoveToLocationCommand(ILogger logger, IDrivetrainSubsystem subsystem, RobotConfig config, Pose2d position) {
```
Comment (dotnetprofessional) [inline code/src/main/java/frc/robot/commands/MoveToLocationCommand.java]: the literals are not going to work well with an IoC. The pattern would be to have these added as properties. Then what ever wanted this class would set the properties at the point of scheduling.

File: code/src/main/java/frc/robot/RobotConfig.java [hunk 3]
```diff
@@ -42,6 +44,8 @@
         public final int rearRightAbsoluteEncoder = 2;
         public final double swerveX = .67 / 2;
         public final double swerveY = .25;
+        public final SwerveDriveKinematics kinematics = new SwerveDriveKinematics(new Translation2d(swerveX, swerveY), new Translation2d(swerveX, -swerveY),
```
Comment (dotnetprofessional) [inline code/src/main/java/frc/robot/RobotConfig.java]: any reason for not just having this code in the SwerveDrivetrain class? It doesn't seem to be used anywhere else and this isn't  a constant in terms of a simple value.

File: code/src/main/java/frc/robot/commands/VisionAdjustCommand.java [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+    public void initialize() {
+
+    }
+
+    public void execute() {
+
+        // network table values
+        NetworkTableEntry tx = m_table.getEntry("tx");
+        NetworkTableEntry ty = m_table.getEntry("ty");
+        NetworkTableEntry ta = m_table.getEntry("ta");
```
Comment (dotnetprofessional) [inline code/src/main/java/frc/robot/commands/VisionAdjustCommand.java]: the calls to NetworkTables would be better behind a class ie LimelightSensor or something. This makes it easier to mock this out. It also means the data can be used in more than one place.

Actually making it a subsystem will make it easier to keep polling the values.

File: code/src/main/java/frc/robot/commands/turnToHeading.java [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+
+import edu.wpi.first.wpilibj.geometry.Pose2d;
+import edu.wpi.first.wpilibj.geometry.Rotation2d;
+import edu.wpi.first.wpilibj.geometry.Translation2d;
+import frc.robot.RobotConfig;
+import frc.robot.common.IDrivetrainSubsystem;
+import frc.robot.common.ILogger;
+import frc.robot.common.TraceableCommand;
+
+public class turnToHeading extends TraceableCommand {
```
Comment (dotnetprofessional) [inline code/src/main/java/frc/robot/commands/turnToHeading.java]: nit: Class names should start with a captial

File: code/src/main/java/frc/robot/commands/turnToHeading.java [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+    private Pose2d[] path;
+    private IDrivetrainSubsystem m_subsystem;
+    private OdometryCommand m_command;
+
+    @Inject
+    public turnToHeading(ILogger logger, IDrivetrainSubsystem subsystem, RobotConfig config, double heading) {
+        super(logger);
+        addRequirements(m_subsystem);
+        m_subsystem = subsystem;
+        path[0] = new Pose2d(m_subsystem.getCurrentPosition().getTranslation(), new Rotation2d(heading));
```
Comment (dotnetprofessional) [inline code/src/main/java/frc/robot/commands/turnToHeading.java]: I think Location would be a better choice as its tracking x,y

File: code/src/test/java/frc/robot/unit/subsystems/drivetrain/Feature_move_robot_using_high_level_commands.java [hunk 7]
```diff
@@ -54,7 +54,7 @@ public void Scenario_rotate_robot_by_x_degrees(final double initialHeading, fina
             final double finalHeading) {
 
         _location.updateHeading(initialHeading);
-        _drivetrain.rotate(rotation);
+        // _drivetrain.rotate(rotation);
```
Comment (dotnetprofessional) [inline code/src/test/java/frc/robot/unit/subsystems/drivetrain/Feature_move_robot_using_high_level_commands.java]: so can I assume all the tests are broken??

**Review by dotnetprofessional — CHANGES_REQUESTED** [review]

**Pull request closed without merging**